### PR TITLE
Change MemoryStorage design & Add opsReqOpts in MongoAutoscaler

### PR DESCRIPTION
**Pull request by ArnobKumarSaha**
Signed-off-by: Arnob kumar saha <arnob@appscode.com>

### CODE REVIEW

File: apis/autoscaling/v1alpha1/constants.go [hunk 1]
```diff
@@ -58,6 +58,11 @@ const (
 	CreateOpsRequest = "CreateOpsRequest"
 )
 
+const (
+	ApplyOptionsIfReady = "ApplyIfReady"
+	ApplyOptionsAlways  = "ApplyAlways"
```
Comment (faem) [inline apis/autoscaling/v1alpha1/constants.go:63]: ApplyOptionsIfReady ApplyOptions = "ApplyIfReady"
ApplyOptionsAlways ApplyOptions = "ApplyAlways"

**Review by faem — CHANGES_REQUESTED** [review]

File: apis/autoscaling/v1alpha1/mongodb_types.go [hunk 2]
```diff
@@ -58,6 +60,9 @@ type MongoDBAutoscaler struct {
 type MongoDBAutoscalerSpec struct {
 	DatabaseRef *core.LocalObjectReference `json:"databaseRef"`
 
+	// This field will be used to control the behaviour of ops-manager
+	OpsReqOptions *MongoDBOpsRequestOptions `json:"opsReqOptions,omitempty"`
```
Comment (tamalsaha) [inline apis/autoscaling/v1alpha1/mongodb_types.go:64]: We should use full words `opsRequestOptions`

File: apis/autoscaling/v1alpha1/type.go [hunk 3]
```diff
@@ -67,15 +67,22 @@ type ComputeAutoscalerSpec struct {
...
-	UsageThreshold int32 `json:"usageThreshold,omitempty"`
+	UsageThresholdPercentage int32 `json:"usageThresholdPercentage,omitempty"`
 
-	// For InMemory storageType, if db uses more than UsageThreshold percentage
+	// For InMemory storageType, if db uses more than UsageThresholdPercentage
 	// of the total memory() `inMemorySizeGB` should be increased by ScalingThreshold percent
+	// Default is 50%
 	// +optional
-	ScalingThreshold int32 `json:"scalingThreshold,omitempty"`
+	ScalingThresholdPercentage int32 `json:"scalingThresholdPercentage,omitempty"`
```
Comment (tamalsaha) [inline apis/autoscaling/v1alpha1/type.go:85]: `scalingFactorPercentage` ?

File: apis/ops/v1alpha1/constant.go [hunk 4]
```diff
@@ -56,3 +56,8 @@ const (
 	FailedToIssueCertificates              = "FailedToIssueCertificates"
 	SuccessfullyReconfiguredTLS            = "SuccessfullyReconfiguredTLS"
 )
+
+const (
+	ApplyOptionsIfReady ApplyOptions = "ApplyIfReady"
+	ApplyOptionsAlways  ApplyOptions = "ApplyAlways"
+)
```
Comment (tamalsaha) [inline apis/ops/v1alpha1/constant.go:63]: Needs a proper enum. Also, just use `IfReady` and `Always`

File: apis/ops/v1alpha1/mongodb_ops_types.go [hunk 5]
```diff
@@ -71,10 +71,14 @@ type MongoDBOpsRequestSpec struct {
...
 	// Specifies information necessary for reprovisioning database
 	Reprovision *Reprovision `json:"reprovision,omitempty"`
+
 	// Specifies the Readiness Criteria
 	ReadinessCriteria *MongoDBReplicaReadinessCriteria `json:"readinessCriteria,omitempty"`
 	// Timeout for each step of the ops request in second. If a step doesn't finish within the specified timeout, the ops request will result in failure.
 	Timeout *metav1.Duration `json:"timeout,omitempty"`
+	// ApplyOptions is to control the execution of OpsRequest depending on the database state.
+	// Default is `ApplyIfReady`
+	ApplyOptions *ApplyOptions `json:"applyOptions,omitempty"`
```
Comment (tamalsaha) [inline apis/ops/v1alpha1/mongodb_ops_types.go:81]: `Default is `ApplyIfReady``

This should not be a pointer. Just use default value.

File: apis/ops/v1alpha1/type.go [hunk 6]
```diff
@@ -276,3 +281,7 @@ type TLSSpec struct {
 	// +optional
 	Remove bool `json:"remove,omitempty"`
 }
+
+// +kubebuilder:validation:Enum=ApplyIfReady;ApplyAlways
+// +kubebuilder:default:="ApplyIfReady"
+type ApplyOptions string
```
Comment (tamalsaha) [inline apis/ops/v1alpha1/type.go:287]: The constants should be moved here right after the type declaration.

File: apis/autoscaling/v1alpha1/type.go [hunk 7]
```diff
@@ -67,15 +67,22 @@ type ComputeAutoscalerSpec struct {
 	// +optional
 	PodLifeTimeThreshold metav1.Duration `json:"podLifeTimeThreshold,omitempty"`
 
-	// For InMemory storageType, if db uses more than UsageThreshold percentage of the total memory() ,
+	// Specifies the dbStorage scaling when db data is stored in Memory
+	MemoryStorage *ComputeMemoryStorageSpec `json:"memoryStorage,omitempty"`
```
[the remark this answers is not in the document]
Comment (tamalsaha) [reply apis/autoscaling/v1alpha1/type.go:71]: `inMemoryStorage` sounds good to me.

File: apis/ops/v1alpha1/type.go [hunk 8]
```diff
@@ -282,6 +282,11 @@ type TLSSpec struct {
 	Remove bool `json:"remove,omitempty"`
 }
 
-// +kubebuilder:validation:Enum=ApplyIfReady;ApplyAlways
-// +kubebuilder:default:="ApplyIfReady"
-type ApplyOptions string
+// +kubebuilder:validation:Enum=IfReady;Always
+// +kubebuilder:default:="IfReady"
```
Comment (tamalsaha) [inline apis/ops/v1alpha1/type.go:286]: `// +kubebuilder:default:="IfReady"` has to be at site this type is used.

File: apis/autoscaling/v1alpha1/mongodb_types.go [hunk 9]
```diff
@@ -90,9 +90,9 @@ type MongoDBOpsRequestOptions struct {
 	// Timeout for each step of the ops request in second. If a step doesn't finish within the specified timeout, the ops request will result in failure.
 	Timeout *metav1.Duration `json:"timeout,omitempty"`
 
-	// ApplyOptions is to control the execution of OpsRequest depending on the database state.
-	// Default is `ApplyIfReady`
-	ApplyOptions *opsapi.ApplyOptions `json:"applyOptions,omitempty"`
+	// ApplyOption is to control the execution of OpsRequest depending on the database state.
+	// Default is `IfReady`
+	ApplyOption opsapi.ApplyOption `json:"applyOption,omitempty"`
```
Comment (tamalsaha) [inline apis/autoscaling/v1alpha1/mongodb_types.go:95]: Add `// +kubebuilder:default:="IfReady"` here

```
Apply opsapi.ApplyOption `json:"apply,omitempty"`
```

**Review by tamalsaha — CHANGES_REQUESTED** [review]

**Review by tamalsaha — APPROVED** [review]

File: apis/autoscaling/v1alpha1/type.go [hunk 10]
```diff
@@ -67,15 +67,22 @@ type ComputeAutoscalerSpec struct {
...
+	UsageThresholdPercentage int32 `json:"usageThresholdPercentage,omitempty"`
 
-	// For InMemory storageType, if db uses more than UsageThreshold percentage
-	// of the total memory() `inMemorySizeGB` should be increased by ScalingThreshold percent
+	// For InMemory storageType, if db uses more than UsageThresholdPercentage
+	// of the total memory() `inMemorySizeGB` should be increased by ScalingFactor percent
+	// Default is 50%
 	// +optional
-	ScalingThreshold int32 `json:"scalingThreshold,omitempty"`
+	ScalingFactorPercentage int32 `json:"scalingFactorPercentage,omitempty"`
```
Comment (faem) [inline apis/autoscaling/v1alpha1/type.go:85]: ignore `inMemorySizeGB` in comment

File: apis/autoscaling/v1alpha1/webhook_helpers.go [hunk 11]
```diff
@@ -69,10 +69,14 @@ func setInMemoryDefaults(computeSpec *ComputeAutoscalerSpec, storageEngine dbapi
...
+		// assigning a dummy pointer to set the defaults
+		computeSpec.InMemoryStorage = &ComputeInMemoryStorageSpec{}
 	}
-	if computeSpec.ScalingThreshold == 0 {
-		computeSpec.ScalingThreshold = DefaultInmemoryScalingThreshold
+	if computeSpec.InMemoryStorage.UsageThresholdPercentage == 0 {
+		computeSpec.InMemoryStorage.UsageThresholdPercentage = DefaultInmemoryUsageThreshold
+	}
+	if computeSpec.InMemoryStorage.ScalingFactorPercentage == 0 {
+		computeSpec.InMemoryStorage.ScalingFactorPercentage = DefaultInmemoryScalingThreshold
```
Comment (faem) [inline apis/autoscaling/v1alpha1/webhook_helpers.go:80]: may be rename the constants.

**Review by faem — CHANGES_REQUESTED** [review]

**Review by faem — DISMISSED** [review]

**Review by faem — APPROVED** [review]

**Review by tamalsaha — APPROVED** [review]

**Pull request merged**